apps/service_providers/utils.py

```diff
@@ -1,62 +1,52 @@
+import dataclasses
 import functools
 from enum import Enum
 
 from django import forms
+from django.db import models
 from django.db.models import Field
+from django_tables2 import tables
 
 from apps.generics.type_select_form import TypeSelectForm
 
 from . import const
 from .models import LlmProvider, LlmProviderType, VoiceProvider, VoiceProviderType
-from .tables import LlmProviderTable, VoiceProviderTable
+from .tables import make_table
 
 
-class ServiceProvider(Enum):
-    llm = const.LLM
-    voice = const.VOICE
+@dataclasses.dataclass
+class ServiceProviderType:
+    slug: str
+    label: str
+    model: models.Model
 
-    @property
-    def model(self):
-        match self:
-            case ServiceProvider.llm:
-                return LlmProvider
-            case ServiceProvider.voice:
-                return VoiceProvider
-        raise ValueError(f"Invalid provider type: {self}")
+    """
+    Enum for the subtypes of this provider type.
+    It is required that the enum has a `form_cls` property which returns
+    the config form class for that subtype.
+    """
+    subtype: Enum
 
-    @property
-    def table(self):
-        match self:
-            case ServiceProvider.llm:
-                return LlmProviderTable
-            case ServiceProvider.voice:
-                return VoiceProviderTable
-        raise ValueError(f"Invalid provider type: {self}")
+
+class ServiceProvider(ServiceProviderType, Enum):
```

Comment:
I did not know python can do this..

Reply:
I didn't either until I was reading the docs